src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -1672,6 +1629,86 @@ static void highly_optimized_short_cases(StrIntrinsicNode::ArgEncoding ae, Regis
   __ jmpb(L_out);
 }
 
+
+
+// Copy the small (< 32 byte) haystack to the stack.  Allows for vector reads without page fault
+// Only done for small haystacks
+// NOTE: This code assumes that the haystack points to a java array type AND there are
+//       at least 8 bytes of header preceeding the haystack pointer.
+// We're copying up to 7 bytes of the header onto the stack along with the haystack bytes.
+// After the copy completes, we adjust the haystack pointer
+// to the valid haystack bytes on the stack.
+//
+// Copy haystack array elements to stack at region
+// (COPIED_HAYSTACK_STACK_OFFSET - COPIED_HAYSTACK_STACK_OFFSET+63) with the following conditions:
+//   It may copy up to 7 bytes that precede the array
+//   It doesn't read beyond the end of the array
+//   There are atleast 31 bytes of stack region beyond the end of array
+// Inputs:
+//   haystack - Address of haystack
+//   haystack_len - Number of elements in haystack
+//   isU - Boolean indicating if each element is Latin1 or UTF16
+//   tmp, xtmp - Scratch registers
+// Output:
+//   haystack - Address of copied string on stack
+
+static void copy_to_stack(Register haystack, Register haystack_len, bool isU,
+                          Register tmp, XMMRegister xtmp, MacroAssembler *_masm) {
+  Label L_moreThan8, L_moreThan16, L_moreThan24, L_adjustHaystack;
+
+  assert(arrayOopDesc::base_offset_in_bytes(isU ? T_CHAR : T_BYTE) >= 8,
```

Comment:
If we had to also optimize for header-size 16, it might be possible to remove one jump here. Looks correct for either size.

Reply:
Yeah. The old code optimized for header-size >= 16. But given that compact headers will soon become the default, I don't think it's worth optimizing for the old header layout.